src/core/components/pin/gc.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+const promisify = require('promisify-es6')
+const CID = require('cids')
+const base32 = require('base32.js')
+const parallel = require('async/parallel')
+const mapLimit = require('async/mapLimit')
+const { Key } = require('interface-datastore')
+const log = require('debug')('ipfs:gc')
+
+// Limit on the number of parallel block remove operations
+const BLOCK_RM_CONCURRENCY = 256
+const MFS_ROOT_DS_KEY = new Key('/local/filesroot')
+
+// Perform mark and sweep garbage collection
+module.exports = function gc (self) {
+  return promisify((callback) => {
+    const start = Date.now()
+    log(`Creating set of marked blocks`)
+
+    self._gcLock.writeLock((lockCb) => {
+      parallel([
+        // Get all blocks from the blockstore
+        (cb) => self._repo.blocks.query({ keysOnly: true }, cb),
+        // Mark all blocks that are being used
+        (cb) => createMarkedSet(self, cb)
+      ], (err, [blocks, markedSet]) => {
+        if (err) {
+          log(`Error - ${err.message}`)
```

Comment:
Please log the full error object and add an appropriate message - this is so we get the stack trace - it's no use if someone turns on `DEBUG` and is not able to debug due to no stack.

```suggestion
          log(`Failed to fetch all block keys and create marked set`, err)
```

Please also fix other instances!